File: backend/app/schemas/subscription.py
"""
Subscription schemas for premium content access system
"""
from pydantic import BaseModel, Field, validator
from typing import Optional, Literal
from datetime import datetime
from uuid import UUID
from enum import Enum


class SubscriptionTier(str, Enum):
    """Subscription tier enumeration"""
    FREE = "free"
    PREMIUM = "premium"


class SubscriptionStatus(str, Enum):
    """Subscription status enumeration"""
    ACTIVE = "active"
    EXPIRED = "expired"
    CANCELLED = "cancelled"
    TRIAL = "trial"


class SubscriptionBase(BaseModel):
    """Base subscription model"""
    tier: SubscriptionTier
    status: SubscriptionStatus
    start_date: datetime
    end_date: datetime
    auto_renew: bool = True
    
    @validator('end_date')
    def validate_end_date(cls, v, values):
        """Ensure end_date is after start_date"""
        if 'start_date' in values and v <= values['start_date']:
            raise ValueError('end_date must be after start_date')
        return v


class SubscriptionCreate(SubscriptionBase):
    """Schema for creating a new subscription"""
    user_id: UUID
    payment_provider: Optional[str] = None
    payment_id: Optional[str] = None
    payment_amount: Optional[float] = Field(None, gt=0)
    payment_currency: str = "UAH"


class Subscription(SubscriptionBase):
    """Complete subscription model"""
    id: UUID
    user_id: UUID
    payment_provider: Optional[str] = None
    payment_id: Optional[str] = None
    payment_amount: Optional[float] = None
    payment_currency: str = "UAH"
    cancelled_at: Optional[datetime] = None
    cancellation_reason: Optional[str] = None
    created_at: datetime
    updated_at: datetime
    
    class Config:
        from_attributes = True


class UserSubscriptionInfo(BaseModel):
    """User subscription information (embedded in user model)"""
    tier: SubscriptionTier = SubscriptionTier.FREE
    status: SubscriptionStatus = SubscriptionStatus.ACTIVE
    start_date: Optional[datetime] = None
    end_date: Optional[datetime] = None
    is_active: bool = False
    
    @validator('is_active', always=True)
    def calculate_is_active(cls, v, values):
        """Calculate if subscription is currently active"""
        if values.get('tier') == SubscriptionTier.FREE:
            return True  # Free tier is always "active"
        
        if values.get('status') != SubscriptionStatus.ACTIVE:
            return False
        
        # Check if subscription has expired
        end_date = values.get('end_date')
        if end_date and end_date < datetime.utcnow():
            return False
        
        return True
    
    class Config:
        from_attributes = True


class SubscriptionStatusResponse(BaseModel):
    """Response model for subscription status check"""
    user_id: UUID
    subscription: UserSubscriptionInfo
    has_premium_access: bool
    features: dict = Field(default_factory=dict)
    
    @validator('features', always=True)
    def set_features(cls, v, values):
        """Set available features based on subscription tier"""
        subscription = values.get('subscription')
        if not subscription:
            return v
        
        is_premium = subscription.tier == SubscriptionTier.PREMIUM and subscription.is_active
        
        return {
            "ai_recipe_generation": is_premium,
            "ai_cooking_tips": is_premium,
            "ai_substitutions": is_premium,
            "ai_nutrition_analysis": is_premium,
            "premium_recipes": is_premium,
            "advanced_search": is_premium,
            "basic_recipes": True,  # Always available
            "basic_search": True,   # Always available
            "favorites": True,      # Always available
        }


class SubscriptionUpdateRequest(BaseModel):
    """Request to update subscription (for admin/testing)"""
    tier: Optional[SubscriptionTier] = None
    status: Optional[SubscriptionStatus] = None
    end_date: Optional[datetime] = None


class PremiumAccessCheck(BaseModel):
    """Response for premium access validation"""
    has_access: bool
    tier: SubscriptionTier
    reason: Optional[str] = None
    upgrade_required: bool = False
    
    @validator('upgrade_required', always=True)
    def set_upgrade_required(cls, v, values):
        """Set upgrade_required based on has_access"""
        return not values.get('has_access', False)


class SubscriptionEvent(BaseModel):
    """Subscription event for audit log"""
    id: UUID
    user_id: UUID
    subscription_id: Optional[UUID] = None
    event_type: str
    old_tier: Optional[SubscriptionTier] = None
    new_tier: Optional[SubscriptionTier] = None
    old_status: Optional[SubscriptionStatus] = None
    new_status: Optional[SubscriptionStatus] = None
    triggered_by: str
    metadata: dict = Field(default_factory=dict)
    created_at: datetime
    
    class Config:
        from_attributes = True


class SubscriptionFeatures(BaseModel):
    """Available features for each subscription tier"""
    tier: SubscriptionTier
    features: dict
    
    @staticmethod
    def get_features(tier: SubscriptionTier) -> dict:
        """Get features for a specific tier"""
        if tier == SubscriptionTier.PREMIUM:
            return {
                "ai_recipe_generation": True,
                "ai_cooking_tips": True,
                "ai_substitutions": True,
                "ai_nutrition_analysis": True,
                "premium_recipes": True,
                "advanced_search": True,
                "basic_recipes": True,
                "basic_search": True,
                "favorites": True,
                "max_favorites": None,  # Unlimited
            }
        else:  # FREE
            return {
                "ai_recipe_generation": False,
                "ai_cooking_tips": False,
                "ai_substitutions": False,
                "ai_nutrition_analysis": False,
                "premium_recipes": False,
                "advanced_search": False,
                "basic_recipes": True,
                "basic_search": True,
                "favorites": True,
                "max_favorites": 50,  # Limited for free tier
            }


class UpgradePrompt(BaseModel):
    """Upgrade prompt information for frontend"""
    title: str = "Upgrade to Premium"
    message: str
    features: list[str]
    cta_text: str = "Upgrade Now"
    
    @staticmethod
    def get_ai_feature_prompt() -> "UpgradePrompt":
        """Get upgrade prompt for AI features"""
        return UpgradePrompt(
            message="AI-powered recipe generation is a premium feature",
            features=[
                "Generate unlimited recipes with AI",
                "Get personalized cooking tips",
                "Smart ingredient substitutions",
                "Detailed nutrition analysis",
                "Access to 500+ premium recipes"
            ]
        )
    
    @staticmethod
    def get_premium_recipe_prompt() -> "UpgradePrompt":
        """Get upgrade prompt for premium recipes"""
        return UpgradePrompt(
            message="This is a premium recipe from professional chefs",
            features=[
                "Access 500+ premium recipes",
                "Exclusive content from verified chefs",
                "AI-powered cooking assistant",
                "Advanced search and filtering",
                "Unlimited favorites"
            ]
        )
    
    @staticmethod
    def get_advanced_search_prompt() -> "UpgradePrompt":
        """Get upgrade prompt for advanced search"""
        return UpgradePrompt(
            message="Search across all recipes with premium access",
            features=[
                "Search premium and basic recipes",
                "Advanced filtering options",
                "AI-powered recipe suggestions",
                "Save unlimited favorites",
                "Access exclusive chef content"
            ]
        )

File: frontend/lib/features/subscription/models/subscription.dart
import 'package:equatable/equatable.dart';

/// Subscription tier enumeration
enum SubscriptionTier {
  free,
  premium;

  String get displayName {
    switch (this) {
      case SubscriptionTier.free:
        return 'Free';
      case SubscriptionTier.premium:
        return 'Premium';
    }
  }

  static SubscriptionTier fromString(String value) {
    switch (value.toLowerCase()) {
      case 'premium':
        return SubscriptionTier.premium;
      case 'free':
      default:
        return SubscriptionTier.free;
    }
  }
}

/// Subscription status enumeration
enum SubscriptionStatus {
  active,
  expired,
  cancelled,
  trial;

  String get displayName {
    switch (this) {
      case SubscriptionStatus.active:
        return 'Active';
      case SubscriptionStatus.expired:
        return 'Expired';
      case SubscriptionStatus.cancelled:
        return 'Cancelled';
      case SubscriptionStatus.trial:
        return 'Trial';
    }
  }

  static SubscriptionStatus fromString(String value) {
    switch (value.toLowerCase()) {
      case 'active':
        return SubscriptionStatus.active;
      case 'expired':
        return SubscriptionStatus.expired;
      case 'cancelled':
        return SubscriptionStatus.cancelled;
      case 'trial':
        return SubscriptionStatus.trial;
      default:
        return SubscriptionStatus.active;
    }
  }
}

/// User subscription information
class UserSubscriptionInfo extends Equatable {
  final SubscriptionTier tier;
  final SubscriptionStatus status;
  final DateTime? startDate;
  final DateTime? endDate;
  final bool isActive;

  const UserSubscriptionInfo({
    required this.tier,
    required this.status,
    this.startDate,
    this.endDate,
    required this.isActive,
  });

  factory UserSubscriptionInfo.fromJson(Map<String, dynamic> json) {
    return UserSubscriptionInfo(
      tier: SubscriptionTier.fromString(json['tier'] ?? 'free'),
      status: SubscriptionStatus.fromString(json['status'] ?? 'active'),
      startDate: json['start_date'] != null
          ? DateTime.tryParse(json['start_date'])
          : null,
      endDate:
          json['end_date'] != null ? DateTime.tryParse(json['end_date']) : null,
      isActive: json['is_active'] == true,
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'tier': tier.name,
      'status': status.name,
      'start_date': startDate?.toIso8601String(),
      'end_date': endDate?.toIso8601String(),
      'is_active': isActive,
    };
  }

  bool get isPremium => tier == SubscriptionTier.premium && isActive;
  bool get isFree => tier == SubscriptionTier.free;

  @override
  List<Object?> get props => [tier, status, startDate, endDate, isActive];
}

/// Subscription features
class SubscriptionFeatures extends Equatable {
  final bool aiRecipeGeneration;
  final bool aiCookingTips;
  final bool aiSubstitutions;
  final bool aiNutritionAnalysis;
  final bool premiumRecipes;
  final bool advancedSearch;
  final bool basicRecipes;
  final bool basicSearch;
  final bool favorites;

  const SubscriptionFeatures({
    required this.aiRecipeGeneration,
    required this.aiCookingTips,
    required this.aiSubstitutions,
    required this.aiNutritionAnalysis,
    required this.premiumRecipes,
    required this.advancedSearch,
    required this.basicRecipes,
    required this.basicSearch,
    required this.favorites,
  });

  factory SubscriptionFeatures.fromJson(Map<String, dynamic> json) {
    return SubscriptionFeatures(
      aiRecipeGeneration: json['ai_recipe_generation'] == true,
      aiCookingTips: json['ai_cooking_tips'] == true,
      aiSubstitutions: json['ai_substitutions'] == true,
      aiNutritionAnalysis: json['ai_nutrition_analysis'] == true,
      premiumRecipes: json['premium_recipes'] == true,
      advancedSearch: json['advanced_search'] == true,
      basicRecipes: json['basic_recipes'] == true,
      basicSearch: json['basic_search'] == true,
      favorites: json['favorites'] == true,
    );
  }

  factory SubscriptionFeatures.free() {
    return const SubscriptionFeatures(
      aiRecipeGeneration: false,
      aiCookingTips: false,
      aiSubstitutions: false,
      aiNutritionAnalysis: false,
      premiumRecipes: false,
      advancedSearch: false,
      basicRecipes: true,
      basicSearch: true,
      favorites: true,
    );
  }

  factory SubscriptionFeatures.premium() {
    return const SubscriptionFeatures(
      aiRecipeGeneration: true,
      aiCookingTips: true,
      aiSubstitutions: true,
      aiNutritionAnalysis: true,
      premiumRecipes: true,
      advancedSearch: true,
      basicRecipes: true,
      basicSearch: true,
      favorites: true,
    );
  }

  @override
  List<Object?> get props => [
        aiRecipeGeneration,
        aiCookingTips,
        aiSubstitutions,
        aiNutritionAnalysis,
        premiumRecipes,
        advancedSearch,
        basicRecipes,
        basicSearch,
        favorites,
      ];
}

/// Complete subscription status response
class SubscriptionStatusResponse extends Equatable {
  final String userId;
  final UserSubscriptionInfo subscription;
  final bool hasPremiumAccess;
  final SubscriptionFeatures features;

  const SubscriptionStatusResponse({
    required this.userId,
    required this.subscription,
    required this.hasPremiumAccess,
    required this.features,
  });

  factory SubscriptionStatusResponse.fromJson(Map<String, dynamic> json) {
    return SubscriptionStatusResponse(
      userId: json['user_id']?.toString() ?? '',
      subscription: UserSubscriptionInfo.fromJson(json['subscription'] ?? {}),
      hasPremiumAccess: json['has_premium_access'] == true,
      features: SubscriptionFeatures.fromJson(json['features'] ?? {}),
    );
  }

  @override
  List<Object?> get props => [userId, subscription, hasPremiumAccess, features];
}

/// Upgrade prompt for premium features
class UpgradePrompt extends Equatable {
  final String title;
  final String message;
  final List<String> features;
  final String ctaText;
  final String ctaAction;

  const UpgradePrompt({
    required this.title,
    required this.message,
    required this.features,
    required this.ctaText,
    required this.ctaAction,
  });

  factory UpgradePrompt.fromJson(Map<String, dynamic> json) {
    return UpgradePrompt(
      title: json['title']?.toString() ?? 'Upgrade to Premium',
      message: json['message']?.toString() ?? '',
      features: (json['features'] as List<dynamic>?)
              ?.map((e) => e.toString())
              .toList() ??
          [],
      ctaText: json['cta_text']?.toString() ?? 'Upgrade Now',
      ctaAction: json['cta_action']?.toString() ?? 'navigate_to_subscription',
    );
  }

  @override
  List<Object?> get props => [title, message, features, ctaText, ctaAction];
}

/// Premium access check result
class PremiumAccessCheck extends Equatable {
  final bool hasAccess;
  final SubscriptionTier tier;
  final String? reason;
  final bool upgradeRequired;

  const PremiumAccessCheck({
    required this.hasAccess,
    required this.tier,
    this.reason,
    required this.upgradeRequired,
  });

  factory PremiumAccessCheck.fromJson(Map<String, dynamic> json) {
    return PremiumAccessCheck(
      hasAccess: json['has_access'] == true,
      tier: SubscriptionTier.fromString(json['tier'] ?? 'free'),
      reason: json['reason']?.toString(),
      upgradeRequired: json['upgrade_required'] == true,
    );
  }

  @override
  List<Object?> get props => [hasAccess, tier, reason, upgradeRequired];
}
